### PR TITLE
Add an option on getBlock to filter vote transactions

### PR DESCRIPTION
**Pull request by AurelienFT**
#### Problem
If we want to use `getBlock` to get only the changes on the blockchain data we have all the vote transaction rely to the consensus. For a use case where we just want to have the changes on the blockchain and don't need the vote transactions those transaction make the response bigger for no reasons.

#### Summary of Changes
Following the `rewards` option I have added an option to filter the simple votes transactions which is `votes`.

Is that possible to also backport it to 1.8 and 1.9 to have the change effective in the next update ? I can right the same changes in the others versions if needed.

### CODE REVIEW

**Comment by t-nelson** [comment]
I'm pretty hesitant to add server-side filtering to `getBlock`. It's already one of the most expensive RPC calls.  Adding a path that deserializes every transaction (potentially twice, depending on output encoding) seems a little misguided.  Perhaps this could be more palatable if restricted to the `json` and `jsonParsed` output modes, where we already have deserialized transactions. 

Ideally I think this would be better off implemented as an AccountsDB Plugin (impending rename!) where it will have lower impact on validator performance

**Comment by AurelienFT** [comment]
> I'm pretty hesitant to add server-side filtering to `getBlock`. It's already one of the most expensive RPC calls. Adding a path that deserializes every transaction (potentially twice, depending on output encoding) seems a little misguided. Perhaps this could be more palatable if restricted to the `json` and `jsonParsed` output modes, where we already have deserialized transactions.
> 
> Ideally I think this would be better off implemented as an AccountsDB Plugin (impending rename!) where it will have lower impact on validator performance

Thank you for feedback. I will restrict it to json and jsonParsed and make the test when they are already deserialize. It will be not a lot of cost has it will just be checking.

**Comment by AurelienFT** [comment]
@t-nelson I have changed the implementation and documentation.
Also if you have an idea of why I don't pass CI it could be cool because I just see a compile error at linking but nothing very verbose.

**Comment by t-nelson** [comment]
@CriesofCarrots what say ye?

File: runtime/src/vote_parser.rs [hunk 1]
```diff
@@ -12,7 +12,7 @@ use {
 pub type ParsedVote = (Pubkey, VoteTransaction, Option<Hash>);
 
 // Used for filtering out votes from the transaction log collector
-pub(crate) fn is_simple_vote_transaction(transaction: &SanitizedTransaction) -> bool {
+pub fn is_simple_vote_transaction(transaction: &SanitizedTransaction) -> bool {
```
Comment (CriesofCarrots) [inline runtime/src/vote_parser.rs:15]: Is this change still needed? I don't see your changeset calling this fn.

File: docs/src/developing/clients/jsonrpc-api.md [hunk 2]
```diff
@@ -375,6 +375,7 @@ Returns identity and transaction information about a confirmed block in the ledg
   "jsonParsed" encoding attempts to use program-specific instruction parsers to return more human-readable and explicit data in the `transaction.message.instructions` list. If "jsonParsed" is requested but a parser cannot be found, the instruction falls back to regular JSON encoding (`accounts`, `data`, and `programIdIndex` fields).
   - (optional) `transactionDetails: <string>` - level of transaction detail to return, either "full", "signatures", or "none". If parameter not provided, the default detail level is "full".
   - (optional) `rewards: bool` - whether to populate the `rewards` array. If parameter not provided, the default includes rewards.
+  - (optional) `votes: bool` - whether to return the simple vote transactions. This option only work with an encoding  `jsonParsed`. If parameter not provided, the default includes simple votes transactions.
```
Comment (CriesofCarrots) [inline docs/src/developing/clients/jsonrpc-api.md:378]: Maybe also note that this parameter will be a noop when `transactionDetails: none`?

**Review by CriesofCarrots — COMMENTED** [review]
I'm okay with this.
I think we could actually support other encodings by adding a `encode_with_filter` fn to `trait Encodable` that takes a closure and performs the Vote program-id check before the Transaction is encoded.

But let's start here for now and assess the need.
Couple nits from me.

File: runtime/src/vote_parser.rs (hunk 1 above)
Comment (AurelienFT) [reply runtime/src/vote_parser.rs]: No, I forget to remove it. Corrected.

File: docs/src/developing/clients/jsonrpc-api.md (hunk 2 above)
Comment (AurelienFT) [reply docs/src/developing/clients/jsonrpc-api.md]: I have specified the documentation. Is that ok for you ?

Comment (CriesofCarrots) [reply docs/src/developing/clients/jsonrpc-api.md]: Yep, it looks good

**Comment by MarkAgarunov** [comment]
The `jsonParsed` output method seems to be considerably slower than the already expensive `json`  for `getBlock`. I'm curious how this change would affect performance of the `getBlock` method both with and without votes enabled. If votes are enabled by default, any overhead from `getBlock` with `jsonParsed` would have a significant overall performance impact on validators. Would it be possible to see a benchmark comparing the 3 (current jsonParsed, change with votes enabled, change with votes disabled)?

**Comment by AurelienFT** [comment]
> The `jsonParsed` output method seems to be considerably slower than the already expensive `json` for `getBlock`. I'm curious how this change would affect performance of the `getBlock` method both with and without votes enabled. If votes are enabled by default, any overhead from `getBlock` with `jsonParsed` would have a significant overall performance impact on validators. Would it be possible to see a benchmark comparing the 3 (current jsonParsed, change with votes enabled, change with votes disabled)?

Currently the votes are always returned by default. The goal here is to lower the reponse of the RPC. It will affect performances of the node if the user that request your RPC add `votes: false` to their call. In my opinion there is two case :
- You have a private node and if don't need this and don't want to affect your performances you just don't add the flag to your calls
- You provide a public RPC node. Either you have a good load balancer and you can handle this change otherwise you just don't accept the calls with this flag.

**Comment by t-nelson** [comment]
There is a third case that you're running an RPC service and are responsible for maintaining an SLA to your customers.  @MarkAgarunov falls into this case :wink:

**Comment by AurelienFT** [comment]
On my mind it was the first case and so I don't think it will change anything because see change just explore the structure is the flag is provided or do nothing otherwise. Maybe a benchmark can prove this. How can I perform a benchmark I should use this doc : https://docs.solana.com/cluster/bench-tps#developer-testnet or there is an other way ? @t-nelson 

Also if it cause too much performance impact (which I doubt but let's admit) is that possible that the RPC provider that can't support this charge just disable the flag like the public RPC currently is blocking the endpoint getBlock and getBlocks ?

**Comment by t-nelson** [comment]
It is absolutely a discreet case from the private RPC case.  Private RPC operators typically also control the software calling them and have access restricted their ports.  RPC service operators may pool multiple clients with differing demands depending on the service tier.  I do not think a request for benchmarks is unreasonable.

Bench TPS is not the tool for this job.  @MarkAgarunov could you elaborate a typical access pattern for benchmark?  Would something like batch-requesting the last ten slots for 1000 consecutive slots, monitoring both time to service request and validator distance from the tip be sufficient?

**Comment by AurelienFT** [comment]
> It is absolutely a discreet case from the private RPC case. Private RPC operators typically also control the software calling them and have access restricted their ports. RPC service operators may pool multiple clients with differing demands depending on the service tier. I do not think a request for benchmarks is unreasonable.


My bad wanted to say second one !

**Comment by AurelienFT** [comment]
@MarkAgarunov Can you answer about the details for the benchmark please ?

**Comment by mvines** [comment]
Note that this solution doesn’t really scale.  Remember that the goal is to build enough usage such that the vote transactions become noise in each block

**Comment by tsutsu** [comment]
@t-nelson @MarkAgarunov As for the concerns about added overhead requiring benchmarking — would these concerns still be relevant if the code here were modified such that the logic for checking for votes is only passed through if `show_votes` is true? I.e. if these lines (https://github.com/solana-labs/solana/blob/1efde28b855b63b770a5e4c04f37dc9cd856d3c8/transaction-status/src/lib.rs#L494-L516) were tucked into a branch on `show_votes`, with the existing code on the other arm of the branch. Then the costs of this would _only_ be being paid by callers/RPC-node operators who allow the use of this flag.

**Comment by polevpn** [comment]
solana is designed so terrible,go to the hell
